openadapt/models_tuning/fine_tune_models.py

```diff
@@ -0,0 +1,207 @@
+        score_per_action = 0
+        for ref_action, prediction_action in zip(
+            reference_action_dicts, prediction_action_dicts
+        ):
+            score_per_action += _calculate_similarity_per_action(
+                ref_action, prediction_action
+            )
+        average_score = score_per_action / len(reference_action_dicts)
+        total_score += average_score
+    average_score = total_score / len(dataset)
+    return average_score
+
+
+def convert_recording_to_dataset(recording):
+    """
+    Convert a recording to a dataset.
+    """
+    dataset = []
+    for action_event in recording.processed_action_events:
+        reference_window = action_event.window_event
+        reference_window_dict = deepcopy(
+            {
+                key: val
+                for key, val in utils.row2dict(reference_window, follow=False).items()
+                if val is not None
+                and not key.endswith("timestamp")
+                and not key.endswith("id")
+                # and not isinstance(getattr(models.WindowEvent, key), property)
+            }
+        )
+        if action_event.children:
+            reference_action_dicts = [
+                deepcopy(
+                    {
+                        key: val
+                        for key, val in utils.row2dict(child, follow=False).items()
+                        if val is not None
+                        and not key.endswith("timestamp")
+                        and not key.endswith("id")
+                        and not isinstance(getattr(models.ActionEvent, key), property)
+                    }
+                )
+                for child in action_event.children
+            ]
+        else:
+            reference_action_dicts = [
+                deepcopy(
+                    {
+                        key: val
+                        for key, val in utils.row2dict(
+                            action_event, follow=False
+                        ).items()
+                        if val is not None
+                        and not key.endswith("timestamp")
+                        and not key.endswith("id")
+                        # and not isinstance(getattr(models.ActionEvent, key), property)
+                    }
+                )
+            ]
+        active_window_dict = reference_window_dict
+
+        prompt = {
+            "reference_window_dict": reference_window_dict,
+            "reference_action_dicts": reference_action_dicts,
+            "active_window_dict": active_window_dict,
+        }
+        dataset.append(prompt)
+    return dataset
+
+
+def _euclidean_distance(point1, point2):
+    if len(point1) != 2 or len(point2) != 2:
+        raise ValueError("Both points must be 2D coordinates (x, y)")
+
+    x1, y1 = point1
+    x2, y2 = point2
+    distance = math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)
+    return distance
+
+
+def _calculate_similarity_per_action(ref_action, prediction_action):
```

Comment:
The simple algorithm is to 
- Verify if 2 actions are of the same type
- If they are both button pressing or releasing, the compare the key
- If they are mouse movement, then calculate the Euclidean between the 2 clicked point.

Reply:
The whole point is the compare how similar the reference action to the predicted action.